src/content/blog/2024-12-04-differences-between-linting-and-type-checking.md

```diff
@@ -0,0 +1,274 @@
+These stricter preferences help prevent API designs that lead developers to create many unused parameters.
+
+A more strict ESLint configuration would be able to report on the `_` parameter:
+
+```ts
+/* eslint @typescript-eslint/no-unused-vars: ["error", { "args": "all", "argsIgnorePattern": "" }] */
+
+type Callback = (id: string, message: string) => void;
+
+declare function registerCallback(callback: Callback): void;
+
+// We only want to log message, not id
+registerCallback((_, message) => console.log(message));
+//                ~
+// eslint(@typescript-eslint/no-unused-vars):
+// '_' is declared but never used.
+```
+
+That extra level of configurability provided by the `no-unused-vars` rules can allow them to act as more granularly configurable versions of their equivalent TypeScript compiler options.
+
+> 💡 See [`no-unused-binary-expressions`: From code review nit to ecosystem improvements](/blog/2024/10/code-review-nit-to-ecosystem-improvements) for more areas of code checking with partial overlap between linting and type checking.
+
+## Is a ESLint Still Useful With TypeScript?
+
+**Yes.**
+
+If you are using TypeScript, it is still very useful to use ESLint.
+In fact, ESLint and TypeScript are at their most powerful when used in conjunction with each other.
+
+### ESLint, With Type Information
+
+Traditional lint rules run on a single file at a time and have no knowledge of other files in the project.
+They can't make decisions on files based on the contents of other files.
+
+However, if your project is set up using TypeScript, you can opt into "type checked" lint rules: rules that can pull in type information.
+In doing so, type checked lint rules can make decisions based on other files.
+
+For example, [`@typescript-eslint/no-for-in-array`](https://typescript-eslint.io/rules/no-for-in-array) is able to detect `for...in` loops over values that are array types, even if the values come from other files.
+TypeScript would not report a type error for a `for...in` loop over an array because doing so is technically type-safe and might be what a developer intended.
+A linter, however, could be configured to notice that the developer probably made a mistake and meant to use a `for...of` loop instead:
+
+```ts
+// declare function getArrayOfNames(): string[];
+import { getArrayOfNames } from "./my-names";
+
+for (const name in getArrayOfNames()) {
+    // eslint(@typescript-eslint/no-for-in-array):
+    // For-in loops over arrays skips holes, returns indices as strings,
+    // and may visit the prototype chain or other enumerable properties.
+    // Use a more robust iteration method such as for-of or array.forEach instead.
+    console.log(name);
+}
+```
+
+Augmenting ESLint with information from TypeScript makes for a more powerful set of lint rules.
+See [Typed Linting: The Most Powerful TypeScript Linting Ever](https://typescript-eslint.io/blog/typed-linting) for more details on typed linting with typescript-eslint.
```

Comment:
I think we need to make mention that type-aware linting has performance implications.